### PR TITLE
[15.0][ADD] fieldservice_logbook

### PR DESCRIPTION
**Pull request by yankinmax**
This module allows you to register and view the logbook on Field Service records.

### CODE REVIEW

File: fieldservice_logbook/models/fsm_order.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    def _read_group_logbook_data(self, logbook):
+        return logbook.read_group(
+            [
+                "|",
+                ("location_id", "in", self.ids),
+                ("equipment_id.location_id", "in", self.ids),
+            ],
+            ["location_id"],
+            ["location_id"],
+        )
```
Comment (Camille0907) [inline fieldservice_logbook/models/fsm_order.py]: I don't know well those models, but it looks strange to have exactly the same code than for  `fsm.location`
Is it normal to compare a `location_id` with `self`, which isn't a `fsm.location` ?

File: fieldservice_logbook/models/fsm_order.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        return {
+            db["location_id"][0]: logbook.search(
+                [
+                    "|",
+                    ("location_id", "=", db["location_id"][0]),
+                    ("equipment_id.location_id", "=", db["location_id"][0]),
+                ]
+            )
+            for db in data
+        }
```
Comment (Camille0907) [inline fieldservice_logbook/models/fsm_order.py]: Same question than above :)

File: fieldservice_logbook/models/fsm_order.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    def create(self, vals_list):
+        records = super().create(vals_list)
+        records._create_equipment_order_log()
+        return records
+
+    def write(self, vals):
+        stage_id = vals.get("stage_id")
+        if stage_id:
+            user = self.env.user
+            closed_orders = self.filtered(lambda l: l.stage_id.is_closed)  # noqa: E741
```
Comment (Camille0907) [inline fieldservice_logbook/models/fsm_order.py]: Nitpicking : why not changing the variable name rather than adding the `noqa` ?

File: fieldservice_logbook/models/fsm_order.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            if not user.has_group("fieldservice.group_fsm_manager") and closed_orders:
+                raise ValidationError(
+                    _(
+                        "The FSM Order has been closed, "
+                        "you are not permitted to changes it's stage."
```
Comment (Camille0907) [inline fieldservice_logbook/models/fsm_order.py]:
```suggestion
                        "you are not permitted to changes its stage."
```

File: fieldservice_logbook/tests/test_fsm_equipment_logbook.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                {
+                    "origin_status": self.equipment_stage_1.name,
+                    "note": "Equipment 1",
+                    "res_id": equipment.id,
+                    "equipment_status": self.equipment_stage_1.name,
+                    "location_id": self.test_location.id,
+                    "equipment_id": equipment.id,
+                    "type": "equipment",
+                    "res_model": "fsm.equipment",
+                    "event_date": fields.Datetime.now(),
```
Comment (Camille0907) [inline fieldservice_logbook/tests/test_fsm_equipment_logbook.py:54]: No need to freeze the time to be sure to have exactly the right value ?

**Review by Camille0907 — APPROVED** [review]
Technically LGTM, just a few questions :)

File: fieldservice_logbook/models/fsm_order.py (hunk 1 above)
Comment (yankinmax) [reply fieldservice_logbook/models/fsm_order.py]: yep, my bad, it should be of course `fsm.order`::`location_id` and `fsm.order`::`equipment_id.location_id`

Comment (yankinmax) [reply fieldservice_logbook/models/fsm_order.py]: `fsm.order` logs are searched in database by `location_id` and `type=order`

**Comment by yankinmax** [comment]
@Camille0907 thanks for the review, I've checked and updated

File: fieldservice/views/fsm_location.xml [hunk 6]
```diff
@@ -102,11 +102,7 @@
                         <group id="main-left">
                             <field name="id" invisible="1" />
                             <field name="fsm_parent_id" domain="[('id', '!=', id)]" />
-                            <field
-                                name="partner_id"
-                                groups="base.group_no_one"
-                                readonly="1"
-                            />
+                            <field name="partner_id" groups="base.group_no_one" />
```
Comment (max3903) [inline fieldservice/views/fsm_location.xml]: :-1: 

This should go into a separate PR. `partner_id` is also the field used to inherit. Same as res.partner and res.users, it should not be editable by the user.

**Comment by max3903** [comment]
@yankinmax Can you provide more details about the business requirement please? I am intereted to understand why the chatter wasn't a good solution. Thank you.

File: fieldservice/views/fsm_location.xml (hunk 6 above)
Comment (yankinmax) [reply fieldservice/views/fsm_location.xml]: Ah sorry, probably bad rebase. Of course, if I would like to change `partner_id` I would inherit the view.

Comment (yankinmax) [reply fieldservice/views/fsm_location.xml]: Good catch indeed.

**Comment by yankinmax** [comment]
Hello @max3903 
For one of the projects, we've used something similar to register and manage FSM records logbook.
And now we have more requests to have kind of generic module to depend only on `fieldservice` with this functionality. This is a good chance to avoid heavily written chatter of the objects itself.
With the usage of a mixin we can simply imply this feature to other objects if needed.

**Comment by yankinmax** [comment]
BTW, failing test has nothing related to either `fieldservice` or `fieldservice_logbook`. FTM I couldn't find where issue comes from.
```
2024-06-21 17:18:01,856 395 ERROR odoo odoo.addons.fieldservice_account_payment.tests.test_fsm_account_payment: ERROR: FSMAccountPaymentCase.test_fsm_account_payment
Traceback (most recent call last):
  File "/__w/field-service/field-service/fieldservice_account_payment/tests/test_fsm_account_payment.py", line 113, in test_fsm_account_payment
    register_payment_action = register_payments.action_create_payments()
  File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 681, in action_create_payments
    payments = self._create_payments()
  File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 675, in _create_payments
    payments = self._init_payments(to_process, edit_mode=edit_mode)
  File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 552, in _init_payments
    payments = self.env['account.payment'].create([x['create_vals'] for x in to_process])
  File "<decorator-gen-198>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/addons/account/models/account_payment.py", line 702, in create
    payments = super().create(vals_list)
  File "<decorator-gen-126>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/addons/mail/models/mail_thread.py", line 265, in create
    threads = super(MailThread, self).create(vals_list)
  File "<decorator-gen-69>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/addons/base/models/ir_fields.py", line 613, in create
    recs = super().create(vals_list)
  File "<decorator-gen-15>", line 2, in create
  File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
    return create(self, arg)
  File "/opt/odoo/odoo/models.py", line 4095, in create
    records = self._create(data_list)
  File "/opt/odoo/odoo/models.py", line 4260, in _create
    records._validate_fields(name for data in data_list for name in data['stored'])
  File "/opt/odoo/odoo/models.py", line 1378, in _validate_fields
    check(self)
  File "/opt/odoo/addons/account/models/account_payment.py", line 648, in _check_payment_method_line_id
    raise ValidationError(_("The selected payment method is not available for this payment, please select the payment method again."))
odoo.exceptions.ValidationError: The selected payment method is not available for this payment, please select the payment method again.
```

**Comment by max3903** [comment]
> Hello @max3903 For one of the projects, we've used something similar to register and manage FSM records logbook. And now we have more requests to have kind of generic module to depend only on `fieldservice` with this functionality. This is a good chance to avoid heavily written chatter of the objects itself. With the usage of a mixin we can simply imply this feature to other objects if needed.

What is the difference between a logbook and a chatter message? 
What can you do with the logbook that you cannot do with the chatter?

**Comment by max3903** [comment]
> BTW, failing test has nothing related to either `fieldservice` or `fieldservice_logbook`. FTM I couldn't find where issue comes from.
> 
> ```
> 2024-06-21 17:18:01,856 395 ERROR odoo odoo.addons.fieldservice_account_payment.tests.test_fsm_account_payment: ERROR: FSMAccountPaymentCase.test_fsm_account_payment
> Traceback (most recent call last):
>   File "/__w/field-service/field-service/fieldservice_account_payment/tests/test_fsm_account_payment.py", line 113, in test_fsm_account_payment
>     register_payment_action = register_payments.action_create_payments()
>   File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 681, in action_create_payments
>     payments = self._create_payments()
>   File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 675, in _create_payments
>     payments = self._init_payments(to_process, edit_mode=edit_mode)
>   File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 552, in _init_payments
>     payments = self.env['account.payment'].create([x['create_vals'] for x in to_process])
>   File "<decorator-gen-198>", line 2, in create
>   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
>     return create(self, arg)
>   File "/opt/odoo/addons/account/models/account_payment.py", line 702, in create
>     payments = super().create(vals_list)
>   File "<decorator-gen-126>", line 2, in create
>   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
>     return create(self, arg)
>   File "/opt/odoo/addons/mail/models/mail_thread.py", line 265, in create
>     threads = super(MailThread, self).create(vals_list)
>   File "<decorator-gen-69>", line 2, in create
>   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
>     return create(self, arg)
>   File "/opt/odoo/odoo/addons/base/models/ir_fields.py", line 613, in create
>     recs = super().create(vals_list)
>   File "<decorator-gen-15>", line 2, in create
>   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
>     return create(self, arg)
>   File "/opt/odoo/odoo/models.py", line 4095, in create
>     records = self._create(data_list)
>   File "/opt/odoo/odoo/models.py", line 4260, in _create
>     records._validate_fields(name for data in data_list for name in data['stored'])
>   File "/opt/odoo/odoo/models.py", line 1378, in _validate_fields
>     check(self)
>   File "/opt/odoo/addons/account/models/account_payment.py", line 648, in _check_payment_method_line_id
>     raise ValidationError(_("The selected payment method is not available for this payment, please select the payment method again."))
> odoo.exceptions.ValidationError: The selected payment method is not available for this payment, please select the payment method again.
> ```

@yankinmax Sounds like an issue with the payment method in the test. This should be fixed in a separate PR. Can you take care of it?

**Comment by yankinmax** [comment]
@max3903 you ask me rather good questions, but I would say just that I'm a developer.
Requests from a customer to approve, refuse, or suggest standards are on the Bussiness Analyst side, so I don't have the right answers for you.
If you're a maintainer of this repository it's of course your decision to take approve this new module or not.
Regarding the issue about payment, I can at least try to fix that in a separate PR.

**Comment by max3903** [comment]
@yankinmax Get the analyst to answer because you will put his answer in the module description. That will help other analysts to decide whether or not they shoud use your module ;)

**Comment by yankinmax** [comment]
> > BTW, failing test has nothing related to either `fieldservice` or `fieldservice_logbook`. FTM I couldn't find where issue comes from.
> > ```
> > 2024-06-21 17:18:01,856 395 ERROR odoo odoo.addons.fieldservice_account_payment.tests.test_fsm_account_payment: ERROR: FSMAccountPaymentCase.test_fsm_account_payment
> > Traceback (most recent call last):
> >   File "/__w/field-service/field-service/fieldservice_account_payment/tests/test_fsm_account_payment.py", line 113, in test_fsm_account_payment
> >     register_payment_action = register_payments.action_create_payments()
> >   File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 681, in action_create_payments
> >     payments = self._create_payments()
> >   File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 675, in _create_payments
> >     payments = self._init_payments(to_process, edit_mode=edit_mode)
> >   File "/opt/odoo/addons/account/wizard/account_payment_register.py", line 552, in _init_payments
> >     payments = self.env['account.payment'].create([x['create_vals'] for x in to_process])
> >   File "<decorator-gen-198>", line 2, in create
> >   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
> >     return create(self, arg)
> >   File "/opt/odoo/addons/account/models/account_payment.py", line 702, in create
> >     payments = super().create(vals_list)
> >   File "<decorator-gen-126>", line 2, in create
> >   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
> >     return create(self, arg)
> >   File "/opt/odoo/addons/mail/models/mail_thread.py", line 265, in create
> >     threads = super(MailThread, self).create(vals_list)
> >   File "<decorator-gen-69>", line 2, in create
> >   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
> >     return create(self, arg)
> >   File "/opt/odoo/odoo/addons/base/models/ir_fields.py", line 613, in create
> >     recs = super().create(vals_list)
> >   File "<decorator-gen-15>", line 2, in create
> >   File "/opt/odoo/odoo/api.py", line 418, in _model_create_multi
> >     return create(self, arg)
> >   File "/opt/odoo/odoo/models.py", line 4095, in create
> >     records = self._create(data_list)
> >   File "/opt/odoo/odoo/models.py", line 4260, in _create
> >     records._validate_fields(name for data in data_list for name in data['stored'])
> >   File "/opt/odoo/odoo/models.py", line 1378, in _validate_fields
> >     check(self)
> >   File "/opt/odoo/addons/account/models/account_payment.py", line 648, in _check_payment_method_line_id
> >     raise ValidationError(_("The selected payment method is not available for this payment, please select the payment method again."))
> > odoo.exceptions.ValidationError: The selected payment method is not available for this payment, please select the payment method again.
> > ```
> 
> @yankinmax Sounds like an issue with the payment method in the test. This should be fixed in a separate PR. Can you take care of it?
@max3903 
FIX PR is here:
- [x] https://github.com/OCA/field-service/pull/1198

**Review by vvrossem — APPROVED** [review]

File: fieldservice_logbook/views/fsm_equipment_logbook.xml [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        <field name="model">fsm.equipment.logbook</field>
+        <field name="arch" type="xml">
+            <form edit="false" delete="false">
+                <sheet>
+                    <group>
+                        <field name="location_id" />
+                        <field name="equipment_id" />
+                        <field name="type" invisible="1" />
+                        <field
+                            name="source"
```
Comment (ivantodorovich) [inline fieldservice_logbook/views/fsm_equipment_logbook.xml:35]: Can't you display the `res_id` (`Many2oneReference`) field directly here, and simply drop this `source` field?

It seems not needed IMO

File: fieldservice_logbook/models/fsm_equipment_logbook_mixin.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        self.ensure_one()
+        return {
+            "origin_status": self.stage_id.name,
+            "note": self.name,
+            "res_id": self.id,
+        }
+
+    @api.model
+    def _create_equipment_log(self):
+        self.env["fsm.equipment.logbook"].create(
```
Comment (ivantodorovich) [inline fieldservice_logbook/models/fsm_equipment_logbook_mixin.py]:
```suggestion
        return self.env["fsm.equipment.logbook"].create(
```

💡 tip: by convenience, to make the method more "overridable"

File: fieldservice_logbook/models/fsm_equipment_logbook_mixin.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            [rec._prepare_log_values() for rec in self]
+        )
+
+    def action_view_equipment_logs(self):
+        self.ensure_one()
```
Comment (ivantodorovich) [inline fieldservice_logbook/models/fsm_equipment_logbook_mixin.py]:
```suggestion
```

💡 tip: Not really needed, as the domain uses the `("id", "in", ...)` form

even if it has no apparent use at the moment

File: fieldservice_logbook/models/fsm_order.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                    }
+                )
+            vals_list.append(values)
+        return vals_list
+
+    def _create_equipment_order_log(self):
+        for record in self:
+            self.env["fsm.equipment.logbook"].create(
+                record._prepare_equipment_log_vals_list()
+            )
```
Comment (ivantodorovich) [inline fieldservice_logbook/models/fsm_order.py]:
```suggestion
        return self.env["fsm.equipment.logbook"].create(
            [
                record._prepare_equipment_log_vals_list()
                for record in self
            ]
        )
```

💡 

1. one single create call
2. return the created recordsets (overridability)

File: fieldservice_logbook/models/fsm_equipment_logbook_mixin.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        "fsm.equipment.logbook", compute="_compute_equipment_logs"
+    )
+
+    def _read_group_logbook_data(self):
+        raise NotImplementedError(
+            "In order to use fsm.equipment.logbook.mixin, "
+            "method _read_group_logbook_data must be implemented"
+        )
+
+    def _map_logbook_data(self):
```
Comment (ivantodorovich) [inline fieldservice_logbook/models/fsm_equipment_logbook_mixin.py]:
```suggestion
    def _map_logbook_data(self, logbook, data):
```

use the expected signature (and perhaps a nice docstring explaining what this method should do)

File: fieldservice_logbook/models/fsm_equipment_logbook_mixin.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+class FsmEquipmentLogbookMixin(models.AbstractModel):
+    _name = "fsm.equipment.logbook.mixin"
+    _description = "Fsm Equipment Logbook Mixin"
+
+    equipment_logs_count = fields.Integer(compute="_compute_equipment_logs")
+    equipment_logs_ids = fields.Many2many(
+        "fsm.equipment.logbook", compute="_compute_equipment_logs"
+    )
+
+    def _read_group_logbook_data(self):
```
Comment (ivantodorovich) [inline fieldservice_logbook/models/fsm_equipment_logbook_mixin.py]:
```suggestion
    def _read_group_logbook_data(self, logbook):
```

use the expected signature (and perhaps a nice docstring explaining what this method should do)

**Review by ivantodorovich — APPROVED** [review]

File: fieldservice_logbook/views/fsm_equipment_logbook.xml (hunk 7 above)
Comment (yankinmax) [reply fieldservice_logbook/views/fsm_equipment_logbook.xml:35]: I've tried and it's completely different from usability point of view. Having just an integer on the tree view and no reference on the form is another functionality.

Comment (ivantodorovich) [reply fieldservice_logbook/views/fsm_equipment_logbook.xml:35]: Oh, I thought the `Many2oneReference` used a similar widget by default.
Nevermind then. Thanks for the reply!
